File: fl_mlkit_translate_text/android/src/main/kotlin/fl/mlkit/translate/text/FlMlKitTranslateTextPlugin.kt
package fl.mlkit.translate.text

import androidx.annotation.NonNull
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.nl.translate.*

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlMlKitTranslateTextPlugin */
class FlMlKitTranslateTextPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel

    private var options: TranslatorOptions? = null
    private var conditions: DownloadConditions? = null
    private var modelManager: RemoteModelManager? = null
    private var translator: Translator? = null
    private var currentSource = TranslateLanguage.ENGLISH
    private var currentTarget = TranslateLanguage.CHINESE


    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "fl_mlkit_translate_text")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "translate" -> {
                val text = call.argument<String>("text")!!
                val downloadModelIfNeeded = call.argument<Boolean>("downloadModelIfNeeded")!!
                if (downloadModelIfNeeded) {
                    getTranslation().downloadModelIfNeeded(getConditions()).addOnSuccessListener {
                        getTranslation().translate(text).addOnSuccessListener { translatedText ->
                            result.success(translatedText)
                        }.addOnFailureListener {
                            result.success(null)
                        }
                    }.addOnFailureListener {
                        result.success(null)
                    }
                } else {
                    getTranslation().translate(text).addOnSuccessListener { translatedText ->
                        result.success(translatedText)
                    }.addOnFailureListener {
                        result.success(null)
                    }
                }
            }
            "switchLanguage" -> {
                val source = call.argument<String>("source")
                val target = call.argument<String>("target")
                dispose()
                currentSource = getTranslateLanguage(source!!)
                currentTarget = getTranslateLanguage(target!!)
                result.success(true)
            }
            "getCurrentLanguage" -> {
                val map = mapOf(
                    "source" to currentSource,
                    "target" to currentTarget,
                )
                result.success(map)
            }
            "getDownloadedModels" -> {
                getModelManager().getDownloadedModels(TranslateRemoteModel::class.java)
                    .addOnSuccessListener { models ->
                        result.success(models.map { model -> model.data })
                    }
                    .addOnFailureListener {
                        result.success(null)
                    }
            }
            "deleteDownloadedModel" -> {
                val model = getTranslateRemoteModel(call.arguments as String)
                getModelManager().deleteDownloadedModel(model).addOnSuccessListener {
                    result.success(true)
                }.addOnFailureListener {
                    result.success(false)
                }
            }
            "downloadedModel" -> {
                val model = getTranslateRemoteModel(call.arguments as String)
                getModelManager().download(model, getConditions()).addOnSuccessListener {
                    result.success(true)
                }.addOnFailureListener {
                    result.success(false)
                }
            }
            "isModelDownloaded" -> {
                val model = getTranslateRemoteModel(call.arguments as String)
                getModelManager().isModelDownloaded(model).addOnSuccessListener { hasModel ->
                    result.success(hasModel)
                }.addOnFailureListener {
                    result.success(false)
                }
            }
            "dispose" -> {
                dispose()
                result.success(true)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun dispose() {
        options = null
        conditions = null
        modelManager = null
        translator?.close()
        translator = null
    }


    private val TranslateRemoteModel.data: Map<String, Any?>
        get() = mapOf(
            "language" to language,
            "isBaseModel" to isBaseModel,
            "modelType" to modelType.name,
        )

    private fun getTranslateLanguage(language: String): String {
        return TranslateLanguage.fromLanguageTag(language)!!
    }

    private fun getTranslateRemoteModel(language: String): TranslateRemoteModel {
        val translateLanguage = getTranslateLanguage(language)
        return TranslateRemoteModel.Builder(translateLanguage).build()
    }

    private fun getConditions(): DownloadConditions {
        if (conditions == null) {
            conditions = DownloadConditions.Builder()
                .requireWifi()
                .build()
        }
        return conditions!!
    }

    private fun getModelManager(): RemoteModelManager {
        if (modelManager == null) {
            modelManager = RemoteModelManager.getInstance()
        }
        return modelManager!!
    }

    private fun getTranslation(): Translator {
        if (translator == null) {
            translator = Translation.getClient(getTranslatorOptions())
        }
        return translator!!
    }

    private fun getTranslatorOptions(): TranslatorOptions {
        if (options == null) {
            options = TranslatorOptions.Builder()
                .setSourceLanguage(currentSource)
                .setTargetLanguage(currentTarget)
                .build()
        }
        return options!!
    }


    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: fl_mlkit_translate_text/ios/Classes/FlMlKitTranslateTextPlugin.swift
import Flutter
import MLKitTranslate
import UIKit

public class FlMlKitTranslateTextPlugin: NSObject, FlutterPlugin {
    private var currentSource = TranslateLanguage.english
    private var currentTarget = TranslateLanguage.chinese
    private var modelManager: ModelManager?
    private var options: TranslatorOptions?
    private var conditions: ModelDownloadConditions?
    private var translator: Translator?

    let fractionCompletedKeyPath: String = "fractionCompleted"
    var downloadProgress: Progress?
    var _result: FlutterResult?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "fl_mlkit_translate_text", binaryMessenger: registrar.messenger())
        let instance = FlMlKitTranslateTextPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "translate":
            let arguments = call.arguments as! [String: Any?]
            let text = arguments["text"] as! String
            let downloadModelIfNeeded = arguments["downloadModelIfNeeded"] as! Bool
            if downloadModelIfNeeded {
                getTranslation().downloadModelIfNeeded(with: getConditions()) { error in
                    if error == nil {
                        self.getTranslation().translate(text) { translatedText, _ in
                            result(translatedText)
                        }
                    } else {
                        result(nil)
                    }
                }
            } else {
                getTranslation().translate(text) { translatedText, _ in
                    result(translatedText)
                }
            }
        case "switchLanguage":
            let arguments = call.arguments as! [String: Any?]
            let source = arguments["source"] as! String
            let target = arguments["target"] as! String
            dispose()
            currentSource = getTranslateLanguage(source)
            currentTarget = getTranslateLanguage(target)
            result(true)
        case "getCurrentLanguage":
            result([
                "source": currentSource,
                "target": currentTarget
            ])
        case "getDownloadedModels":
            let localModels = getModelManager().downloadedTranslateModels
            result(localModels.map {
                $0.data
            })
        case "deleteDownloadedModel":
            getModelManager().deleteDownloadedModel(getTranslateRemoteModel(call)) { error in
                result(error == nil)
            }
        case "downloadedModel":
            _result = result
            downloadProgress?.removeObserver(self, forKeyPath: fractionCompletedKeyPath)
            downloadProgress = getModelManager().download(getTranslateRemoteModel(call), conditions: getConditions())
            downloadProgress!.addObserver(self, forKeyPath: fractionCompletedKeyPath, options: .new, context: nil)
        case "isModelDownloaded":
            result(getModelManager().isModelDownloaded(getTranslateRemoteModel(call)))
        case "dispose":
            dispose()
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    override public func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == fractionCompletedKeyPath, change?[NSKeyValueChangeKey.newKey] as! Int == 1 {
            _result?(true)
        }
    }

    private func dispose() {
        downloadProgress?.removeObserver(self, forKeyPath: fractionCompletedKeyPath)
        downloadProgress = nil
        _result = nil
        options = nil
        conditions = nil
        modelManager = nil
        translator = nil
    }

    private func getConditions() -> ModelDownloadConditions {
        if conditions == nil {
            conditions = ModelDownloadConditions(
                    allowsCellularAccess: false,
                    allowsBackgroundDownloading: true
            )
        }
        return conditions!
    }

    private func getModelManager() -> ModelManager {
        if modelManager == nil {
            modelManager = ModelManager.modelManager()
        }
        return modelManager!
    }

    private func getTranslateLanguage(_ language: String) -> TranslateLanguage {
        return TranslateLanguage(rawValue: language)
    }

    private func getTranslateRemoteModel(_ call: FlutterMethodCall) -> TranslateRemoteModel {
        let language = call.arguments as! String
        let translateLanguage = getTranslateLanguage(language)
        return TranslateRemoteModel.translateRemoteModel(language: translateLanguage)
    }

    private func getTranslation() -> Translator {
        if translator == nil {
            translator = Translator.translator(options: getTranslatorOptions())
        }
        return translator!
    }

    private func getTranslatorOptions() -> TranslatorOptions {
        if options == nil {
            options = TranslatorOptions(sourceLanguage: currentSource, targetLanguage: currentTarget)
        }
        return options!
    }
}

extension TranslateRemoteModel {
    var data: [String: Any?] {
        [
            "language": language
        ]
    }
}
